keras_nlp/models/llama/llama_decoder.py

```diff
@@ -0,0 +1,211 @@
+        self.kernel_initializer = keras.initializers.get(kernel_initializer)
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+
+        self.supports_masking = True
+        self._decoder_sequence_shape = None
+
+    def build(self, decoder_sequence_shape):
+        self._decoder_sequence_shape = decoder_sequence_shape
+        hidden_dim = decoder_sequence_shape[-1]
+        # Self attention layers.
+        self._self_attention_layer = LlamaAttention(
+            num_heads=self.num_heads,
+            hidden_dim=hidden_dim,
+            num_key_value_heads=self.num_key_value_heads,
+            dropout=self.dropout,
+            rotary_percentage=self.rotary_percentage,
+            rotary_max_wavelength=self.rotary_max_wavelength,
+            max_sequence_length=self.max_sequence_length,
+            rope_scaling_type=self.rope_scaling_type,
+            rope_scaling_factor=self.rope_scaling_factor,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+        self._self_attention_layer.build(decoder_sequence_shape)
+
+        self._self_attention_layernorm = LlamaLayerNorm(
+            epsilon=self.layer_norm_epsilon,
+        )
+        self._self_attention_layernorm.build(decoder_sequence_shape)
+
+        self._self_attention_dropout = keras.layers.Dropout(
+            rate=self.dropout,
+        )
+
+        # Feedforward layers.
+        self._feedforward_intermediate_dense = keras.layers.Dense(
+            self.intermediate_dim,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+        self._feedforward_intermediate_dense.build(decoder_sequence_shape)
+
+        self._feedforward_gate_dense = keras.layers.Dense(
+            self.intermediate_dim,
+            activation=self.activation,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+        self._feedforward_gate_dense.build(decoder_sequence_shape)
+
+        self._feedforward_output_dense = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+
+        intermediate_shape = list(decoder_sequence_shape)
+        intermediate_shape[-1] = self.intermediate_dim
+        self._feedforward_output_dense.build(tuple(intermediate_shape))
+
+        self._feedforward_layernorm = LlamaLayerNorm(
+            epsilon=self.layer_norm_epsilon,
+        )
+        self._feedforward_layernorm.build(decoder_sequence_shape)
+
+        self._feedforward_dropout = keras.layers.Dropout(
+            rate=self.dropout,
+        )
+        self.built = True
+
+    def call(
```

Comment:
this will need caching support before we are done here, but ok to add later.

Reply:
cool

Reply:
done!